### PR TITLE
Created UI-tests for PitMutation Plugin

### PR DESCRIPTION
**Pull request by jwalteri**
Within the course "Testing Enterprise Applications" by Prof. Dr. @uhafner from Munich University of Applied Sciences, we created UI tests for the PitMutation plugin.

Please contact Prof. Dr. Hafner (@uhafner) if you have any questions.

The UI tests contain among others the following tests:
- PitMutation Plugin is shown in UI
- All tables were checked for correctness
- All links were checked for usability
- Mutation data is displayed correctly
- Test data is used

The tests were implemented using pair programming.
Authors: Johannes Walter (@jwalteri) & Katharina Winkler  (@kathwin)

### CODE REVIEW

**Comment by jwalteri** [comment]
I would like to politely ask if you can please add @uhafner as a reviewer, as I unfortunately cannot do this myself.

File: ui-tests/src/main/java/org/jenkins/plugins/pitmutation/WebElementUtils.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.openqa.selenium.WebElement;
+
+import java.util.List;
+
+import static org.jenkinsci.test.acceptance.Matchers.by;
+
+/**
+ * Represents utils for WebElements.
+ */
```
Comment (uhafner) [inline ui-tests/src/main/java/org/jenkins/plugins/pitmutation/WebElementUtils.java:11]: I think inlining that class would make the code more readable (at least for someone already familiar with ATH)

File: ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+     */
+    @Test
+    public void verifyTestExamined() {
+        MutationDetailView classView = getClassView();
+
+        List<String> tests = classView.getTestsExamined();
+
+        assertThat(tests.size()).isEqualTo(21);
+        assertThat(tests.get(0)).contains("edu.hm.hafner.util.PathUtilTest", "shouldReturnFallbackOnError",
+            "[test-template-invocation:#3] (8 ms)");
```
Comment (uhafner) [inline ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java:58]: In AssertJ you can also write it like:
```
        assertThat(tests, StringAssert.class)
            .hasSize(21)
            .element(0)
            .contains(("edu.hm.hafner.util.PathUtilTest", "shouldReturnFallbackOnError", "[test-template-invocation:#3] (8 ms)");
```

File: ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    @Test
+    public void verifyMutators() {
+        MutationDetailView classView = getClassView();
+
+        List<String> mutators = classView.getActiveMutators();
+
+        assertThat(mutators.size()).isEqualTo(11);
+        assertThat(mutators.get(0)).isEqualTo("CONDITIONALS_BOUNDARY");
+    }
+
```
Comment (uhafner) [inline ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java:73]: In UI tests it makes more sense to reduce the number of tests (the initial setup is too slow).  So I would recommend to combine those tests with the same setup into a single test.

File: ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    public void verifySource() {
+        MutationDetailView classView = getClassView();
+
+        MutationSourceTable sourceTable = classView.getMutationSource();
+        MutationSourceTableEntry entry = sourceTable.getDataEntries().get(ROW_KEY);
+
+        assertThat(entry.getKey()).isEqualTo(ROW_KEY);
+        assertThat(entry.getSourceCode()).contains("return Files.exists(Paths.get(fileName));");
+        assertThat(entry.getMutationDetail()).contains("edu/hm/hafner/util/PathUtil::exists");
+        assertThat(entry.getClickable().getText()).isEqualTo("2");
```
Comment (uhafner) [inline ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java:122]: The mutation detail is always visible. The use CSS class `pop` indicates that it should show up as a popup. @gigaSproule Is this a known bug of the plugin?

**Review by uhafner — CHANGES_REQUESTED** [review]
In general, a well written PR that provides a lot of new UI tests for the pit-plugin. 

In order to run these tests as part of pull request checks use an action simpler to https://github.com/jenkinsci/code-coverage-api-plugin/blob/master/.github/workflows/ui-tests.yml

File: ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java (hunk 4 above)
Comment (gigaSproule) [reply ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java:122]: I'm not aware of this being a known bug, but if I'm honest, I've not used Jenkins, and by extension this plugin, in a _very_ long time, so I wouldn't commit one way or the other.

Comment (uhafner) [reply ui-tests/src/test/java/org/jenkins/plugins/pitmutation/PitMutationTest.java:122]: @jwalteri Any plans to finish this PR?

**Comment by uhafner** [comment]
@jwalteri Any plans to finish this PR?

**Comment by uhafner** [comment]
I think we can close this issue. I don't think, that someone is planning to finish the open tasks.

BTW: in the meantime I added support for PIT in the [coverage plugin](https://github.com/jenkinsci/code-coverage-api-plugin), so basically the whole functionality of the PIT plugin (and more) is now also available there as well 🚀 

Example pull request with GitHub PIT checks:
- https://github.com/jenkinsci/coverage-model/pull/13/checks?check_run_id=12801605542

Example Jenkins report of PIT in the coverage plugin:
- https://ci.jenkins.io/job/Plugins/job/coverage-model/job/PR-13/2/pit/

**Pull request closed without merging**